cloud-controller-manager/do/firewall_controller.go

```diff
@@ -0,0 +1,304 @@
+	serviceLister corelisters.ServiceLister
+	fwManager     FirewallManager
+	fwCache       firewallCache
+}
+
+// NewFirewallController returns a new firewall controller to reconcile public access firewall state.
+func NewFirewallController(
+	ctx context.Context,
+	kubeClient clientset.Interface,
+	client *godo.Client,
+	serviceInformer coreinformers.ServiceInformer,
+	fwManager FirewallManager,
+	tags []string,
+) *FirewallController {
+	fc := &FirewallController{
+		kubeClient: kubeClient,
+		client:     client,
+		tags:       tags,
+		fwManager:  fwManager,
+	}
+
+	serviceInformer.Informer().AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(cur interface{}) {
+				fc.ensureReconciledFirewall(ctx)
+			},
+			UpdateFunc: func(old, cur interface{}) {
+				fc.ensureReconciledFirewall(ctx)
+			},
+			DeleteFunc: func(obj interface{}) {
+				fc.ensureReconciledFirewall(ctx)
+			},
+		},
+		serviceSyncPeriod,
+	)
+	fc.serviceLister = serviceInformer.Lister()
+
+	return fc
+}
+
+// Run starts the firewall controller loop.
+func (fc *FirewallController) Run(stopCh <-chan struct{}) {
+	ctx, cancel := context.WithTimeout(context.Background(), serviceSyncPeriod)
+	defer cancel()
+
+	wait.Until(func() {
+		currentFirewall, err := fc.fwManager.Get(ctx)
+		if err != nil {
+			klog.Errorf("failed to get worker firewall: %s", err)
+		}
+
+		if currentFirewall != nil {
+			fc.fwCache.mu.RLock()
+			isEqual := cmp.Equal(currentFirewall, fc.fwCache.firewall)
+			fc.fwCache.mu.RUnlock()
+
+			if isEqual {
+				return
+			}
+		}
+		err = fc.ensureReconciledFirewall(ctx)
+		if err != nil {
+			klog.Errorf("failed to reconcile worker firewall: %s", err)
+		}
+
+	}, firewallReconcileRequency, stopCh)
+}
+
+// Get returns the current public access firewall representation.
+func (fm *firewallManagerOp) Get(ctx context.Context) (*godo.Firewall, error) {
+	// check cache and query the API firewall service to get firewall ID, if it exists. Return it. If not, continue.
+	if fm.fwCache.firewall != nil {
+		fm.fwCache.mu.RLock()
+		fwID := fm.fwCache.firewall.ID
+		fm.fwCache.mu.RUnlock()
+
+		fw, resp, err := fm.client.Firewalls.Get(ctx, fwID)
+		if err != nil && (resp == nil || resp.StatusCode != http.StatusNotFound) {
+			return nil, fmt.Errorf("failed to retrieve firewall by ID: %s", err)
+		}
+		if resp.StatusCode == http.StatusNotFound {
+			klog.Warningf("unable to retrieve firewall by ID because it no longer exists")
+		}
+		if fw != nil {
+			return fw, nil
+		}
+	}
+
+	// iterate through firewall API provided list and return the firewall with the matching firewall name.
+	f := func(fw godo.Firewall) bool {
+		return cmp.Equal(fw.Name, fm.workerFirewallName)
+	}
+	klog.Info("filtering firewall list for the firewall that has the expected firewall name")
+	fw, err := filterFirewallList(ctx, fm.client, f)
+	if err != nil {
+		return nil, fmt.Errorf("failed to retrieve list of firewalls from DO API: %s", err)
+	}
+	return fw, nil
```

Comment:
Any reason not to update the cached firewall here, when `fw != nil`?

Reply:
We are just getting the firewall and returning it. The cache gets updated in the `Set()` function.

Reply:
I'm going to update the `Set()` function with better comments.

Reply:
Is `Get` only called by `Set`? My thinking was, if an external caller calls `Get` it may as well update the cache so that future calls to `Get` don't need to fetch the firewall again.

Reply:
Yes it is currently only being called by `Set()`. 

The thinking with `Get()` is that it's main job were these tasks:
- make a Get request to firewall service when a cached firewall ID exists. O(1) operation is ideal, so it's the first attempt at getting the current firewall (current as in the firewall service firewall) and returning it.
- if that fails, try to list all firewalls and get it by firewall name, then return that current firewall.
- if that doesn't work, just return nil.

`Set()`:
- takes in some inbound rules and check if the local cache(if it exists) has the same inbound rules. if it does, return no error.
- if target firewall does exist, then we save it's ID for future update request (to update the inbound rules).
```go
fm.fwCache.mu.RLock()
fwID = targetFirewall.ID
fm.fwCache.mu.RUnlock()
```
- if they don't match and target firewall does not exist, we `Get()` it and save it's ID or create it and save it's ID. The ID is needed for future update request (to update the inbound rules)
```go
fm.fwCache.mu.RLock()
fwID = currentFirewall.ID
fm.fwCache.mu.RUnlock()
```
- make a firewall request object to include the new inbound rules, then pass in this request along with the `fwID` to update the firewall service.
- IFF it succeeds update the cache and return nil otherwise return an error.

Reply:
FYI, thanks to one of your comments I noticed a weird return. So I updated that to match what I describe above. I'll push that up in a min.